File: Minsk/CodeAnalysis/Syntax/LiteralExpressionSyntax.cs
namespace Minsk.CodeAnalysis.Syntax;

public sealed class LiteralExpressionSyntax : ExpressionSyntax
{
    public SyntaxToken LiteralToken { get; }

    public override SyntaxKind Kind => SyntaxKind.LiteralExpression;
    public override IEnumerable<SyntaxNode> Children => new[] { LiteralToken };

    public LiteralExpressionSyntax(SyntaxToken literalToken)
    {
        LiteralToken = literalToken;
    }
}

File: nim/src/minskpkg/codeAnalysis/syntax/expressions/literalExpressionSyntax.nim
import minskpkg/minskObject
import minskpkg/codeAnalysis/syntax/[
  expressionSyntax,
  syntaxKind,
  syntaxNode,
  syntaxToken,
]

type
  LiteralExpressionSyntax* = ref object of ExpressionSyntax
    literalToken*: SyntaxToken
    value*: MinskObject

proc newLiteralExpressionSyntax*(
  literalToken: SyntaxToken,
  value: MinskObject = literalToken.value
): LiteralExpressionSyntax =
  new(result)
  result.literalToken = literalToken
  result.value = value

method kind*(self: LiteralExpressionSyntax): SyntaxKind =
  return SyntaxKind.LiteralExpression

method children*(self: LiteralExpressionSyntax): seq[SyntaxNode] =
  result.add(self.literalToken)
